doc/source/development/rfc/rfc95_standard_int_types.rst

```diff
@@ -0,0 +1,216 @@
+specifically for the CPL_FRMT_GIB/CPL_FRMT_GUIB replacement
+which was harder to automate.
+
+SWIG bindings
+-------------
+
+While .i files are impacted to cope with the C type changes, the language
+specific API of SWIG bindings is not impacted.
+
+Backward compatibility
+----------------------
+
+C and C++ API and ABI are impacted.
+
+Main impacts are:
+
+* GBool was aliased to ``int``: changing to ``bool`` in C++ methods of the
+  OGRStyleXXXX classes affect the API and ABI. Impact should be modest as this
+  functionality is thought to be marginally used, and this only impacts the C++ API
+  (the C API for those methods uses ``int`` and not ``GBool``)
+
+* GIntBig/GInt64 was aliased to ``long long`` and GUIntBig/GUInt64 to
+  ```unsigned long long``. While ``int64_t`` and ``long long`` have in practice
+  same width and signedness, they are formally different data types.
+  For scalar usage of those types, compilers shouldn't warn.
+  But for usage of those types as pointers, compilers warn in C
+  (``-Wincompatible-pointer-types`` with gcc) and error outs in C++
+  (unless ``-fpermissive`` is passed), when mixing a
+  ``int64_t`` pointer and a ``long long`` pointer (similarly for their unsigned
+  counterparts).
+
+  C Raster API impacts (and equivalent C++ methods):
+
+  - :cpp:func:`GDALGetDefaultHistogramEx` (e.g. used by QGIS)
+  - :cpp:func:`GDALSetDefaultHistogramEx`
+  - :cpp:func:`GDALGetRasterHistogramEx` (e.g. used by QGIS)
+  - :cpp:func:`GDALGetVirtualMemAuto`
+
+  C Vector API impacts (and equivalent C++ methods):
+
+  - :cpp:func:`OGR_F_GetFieldAsInteger64List`
+  - :cpp:func:`OGR_F_SetFieldInteger64List`
+
+  C Multidimension API impacts (and equivalent C++ methods):
+
+  - :cpp:func:`GDALGroupCreateDimension`
+  - :cpp:func:`GDALGroupCreateAttribute`
+  - :cpp:func:`GDALMDArrayRead`
+  - :cpp:func:`GDALMDArrayWrite`
+  - :cpp:func:`GDALMDArrayAdviseRead`
+  - :cpp:func:`GDALMDArrayAdviseReadEx`
+  - :cpp:func:`GDALMDArrayCreateAttribute`
+  - :cpp:func:`GDALMDArrayResize`
+  - :cpp:func:`GDALMDArrayGetBlockSize`
+  - :cpp:func:`GDALMDArrayGetStatistics`
+  - :cpp:func:`GDALMDArrayComputeStatistics`
+  - :cpp:func:`GDALMDArrayComputeStatisticsEx`
+  - :cpp:func:`GDALAttributeGetDimensionsSize`
+
+* Out-of-tree drivers are also impacted:
+
+  - the vector ones that implement :cpp:func:`OGRLayer::ISetFeature`,
+    :cpp:func:`OGRLayer::SetNextByIndex`, :cpp:func:`OGRLayer::DeleteFeature`,
+    :cpp:func:`OGRLayer::GetFeatureCount`
+
+  - the ones that implement the multidimensional API (no publicly ones known by us)
+
+Questions to answer before adoption
+-----------------------------------
+
+Q1: What to do with the 64-bit integer changes?
+
+A1: Potential alternatives:
+
+1) Proceed with them, keep GDAL 3.8 as version number. External code will have
+   to use #ifdef if support for GDAL < 3.8 and >= 3.8 is needed.
+
+2) Same as 1), but bump GDAL version number to 4.0. Cleaner way
```

Comment:
I think doing this as a 4.0 change makes the most sense since it is a breaking change requiring changes in upstream codebases.

Reply:
@snowman2 @rouault I mentioned implications for Rasterio and Fiona on gdal-dev and will repeat myself here. I'm not sure what the necessary changes would be for them... they're Cython, not C, and the Cython project feels like conditional compilation directives was a mistake 😂 

It's not clear to me that we can do just this https://github.com/PDAL/PDAL/pull/4179/files#diff-22b766759d40394aa74c390f2ff295b1737020f1a824e103238b60efa9e2814fR36 like for PDAL.

Introducing stubs to support type changes would be a drag, hopefully that could be avoided.